### PR TITLE
Make targets an attribute of compilers

### PR DESCRIPTION
**Pull request by becker33**
Allows users to use `spack compiler find` to add compilers for the same platform/os but different targets when spack is installed on a file system mounted on machines with different targets.

The default behavior of operating systems will be to assume all compilers it can find are for the target on which it is executing. The cray backend operating system will use the keyword 'any', as the cray targeting system is dependent on targeting modules to change the behavior of the same compiler wrapper. No other operating systems currently differ from the default compiler find behavior, so no other changes should be necessary.

### CODE REVIEW

**Comment by becker33** [comment]
Briefly marked WIP because I thought I found a bug. Turns out it was a bug in my workflow, not the code, so I removed the WIP tag.

File: lib/spack/spack/architecture.py [hunk 1]
```diff
@@ -77,6 +77,7 @@
 """
 import os
 import inspect
+import platform as pltfrm
```
Comment (tgamblin) [inline lib/spack/spack/architecture.py]: Make this `python_platform` or something.

File: lib/spack/spack/compilers/__init__.py [hunk 2]
```diff
@@ -220,6 +221,11 @@ def get_compilers(cspec):
             if arch_spec and os != arch_spec.platform_os:
                 continue
 
+            target = items.get('target', None)
```
Comment (tgamblin) [inline lib/spack/spack/compilers/__init__.py]: Add a comment for this block

File: lib/spack/spack/concretize.py [hunk 3]
```diff
@@ -302,7 +302,7 @@ def concretize_compiler(self, spec):
            link to this one, to maximize compatibility.
         """
         # Pass on concretizing the compiler if the target is not yet determined
```
Comment (tgamblin) [inline lib/spack/spack/concretize.py]: update this comment to say we want BOTH OS and target before picking a compiler.

File: lib/spack/spack/concretize.py [hunk 4]
```diff
@@ -367,7 +369,7 @@ def concretize_compiler_flags(self, spec):
         Default specs set at the compiler level will still be added later.
         """
 
-        if not spec.architecture.platform_os:
+        if not (spec.architecture.platform_os and spec.architecture.target):
             # Although this usually means changed, this means awaiting other
             # changes
```
Comment (tgamblin) [inline lib/spack/spack/concretize.py]: Reword: We haven't changed, but other changes need to happen before we continue.  `return True` here to force concretization to keep running.

Comment (tgamblin) [inline lib/spack/spack/concretize.py]: perhaps add "because we don't have either an os or a target"

**Comment by becker33** [comment]
@tgamblin Should be good now.

**Pull request merged**